mm2src/mm2_main/src/lp_swap/taker_swap.rs

```diff
@@ -1943,24 +1987,37 @@ impl TakerSwap {
                 .unwrap_or_else(|| taker_coin.requires_notarization()),
         };
 
+        #[cfg(any(test, feature = "run-docker-tests"))]
+        let fail_at = std::env::var("TAKER_FAIL_AT").map(FailAt::from).ok();
+
         let swap = TakerSwap::new(
-            ctx,
+            ctx.clone(),
             maker,
             data.maker_amount.clone().into(),
             data.taker_amount.clone().into(),
             my_persistent_pub,
             saved.uuid,
             Some(saved.uuid),
             conf_settings,
-            maker_coin,
-            taker_coin,
+            maker_coin.clone(),
+            taker_coin.clone(),
             data.lock_duration,
             data.p2p_privkey.map(SerializableSecp256k1Keypair::into_inner),
+            #[cfg(any(test, feature = "run-docker-tests"))]
+            fail_at,
         );
-        let command = saved.events.last().unwrap().get_command();
-        for saved_event in saved.events {
-            swap.apply_event(saved_event.event);
+        let mut command = saved.events.last().unwrap().get_command();
+        for saved_event in &saved.events {
+            swap.apply_event(saved_event.event.clone());
         }
+
+        if taker_coin.is_supported_by_watchers()
+            && maker_coin.is_supported_by_watchers()
+            && saved.watcher_message_sent()
```

Comment:
How about moving this check inside `check_watcher_payments` or `get_next_command_based_on_watcher_activity` after it's renamed to make this function self contained and to not leak watcher related logic outside it? It will be a negative check instead with early return.

Reply:
It's fine to keep it as is.